WebKitBrowser/InjectedBundle/main.cpp

```diff
@@ -97,28 +109,117 @@ static class PluginHost {
     {
-
         // Whitelist origin/domain pairs for CORS, if set.
         if (_whiteListedOriginDomainPairs) {
             _whiteListedOriginDomainPairs->AddWhiteListToWebKit(bundle);
         }
     }
 
+#ifdef WEBKIT_GLIB_API
+
+private:
+    static void automationMilestone(const char* arg1, const char* arg2, const char* arg3)
+    {
+        g_printerr("TEST TRACE: \"%s\" \"%s\" \"%s\"\n", arg1, arg2, arg3);
+        TRACE_GLOBAL(Trace::Information, (_T("TEST TRACE: \"%s\" \"%s\" \"%s\""), arg1, arg2, arg3));
+    }
+
+    static void windowObjectClearedCallback(WebKitScriptWorld* world, WebKitWebPage*, WebKitFrame* frame)
+    {
+        if (webkit_frame_is_main_frame(frame) == false)
+            return;
+
+        JSCContext* jsContext = webkit_frame_get_js_context_for_script_world(frame, world);
+
+        JSCValue* automation = jsc_value_new_object(jsContext, nullptr, nullptr);
+        JSCValue* function = jsc_value_new_function(jsContext, nullptr, reinterpret_cast<GCallback>(automationMilestone),
+            nullptr, nullptr, G_TYPE_NONE, 3, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_STRING);
+        jsc_value_object_set_property(automation, "Milestone", function);
+        g_object_unref(function);
+        jsc_context_set_value(jsContext, "automation", automation);
+        g_object_unref(automation);
+
+        static const char wpeNotifyWPEFramework[] = "var wpe = {};\n"
+            "wpe.NotifyWPEFramework = function() {\n"
+            "  let retval = new Array;\n"
+            "  for (let i = 0; i < arguments.length; i++) {\n"
+            "    retval[i] = arguments[i];\n"
+            "  }\n"
+            "  window.webkit.messageHandlers.wpeNotifyWPEFramework.postMessage(retval);\n"
+            "}";
+        JSCValue* result = jsc_context_evaluate(jsContext, wpeNotifyWPEFramework, -1);
+        g_object_unref(result);
+
+        g_object_unref(jsContext);
+    }
+
+    static void pageCreatedCallback(WebKitWebExtension*, WebKitWebPage* page, PluginHost* host)
+    {
+        // Enforce the creation of the script world global context in the main frame.
+        JSCContext* jsContext = webkit_frame_get_js_context_for_script_world(webkit_web_page_get_main_frame(page), host->_scriptWorld);
+        g_object_unref(jsContext);
+
+        g_signal_connect(page, "console-message-sent", G_CALLBACK(consoleMessageSentCallback), nullptr);
+    }
+
+    static void consoleMessageSentCallback(WebKitWebPage* page, WebKitConsoleMessage* message)
+    {
+        string messageString = Core::ToString(webkit_console_message_get_text(message));
+
+        const uint16_t maxStringLength = Trace::TRACINGBUFFERSIZE - 1;
+        if (messageString.length() > maxStringLength) {
+            messageString = messageString.substr(0, maxStringLength);
+        }
+
+        // TODO: use "Trace" classes for different levels.
+        TRACE_GLOBAL(Trace::Information, (messageString));
+    }
+
+    WKBundleRef _bundle;
+    WebKitScriptWorld* _scriptWorld;
```

Comment:
This would be better in a separate source file. I'd also suggest sticking with WK C API, or migrate the whole thing to GLib APIs which will be much bigger effort.

Reply:
Unfortunatly we have to support both. Hence why there is yet another Pull Request lined up to fix this. The C-API is the old one (a lot of our customers still use it) The GLIB API is the new one and we are moving towards it. So both need to be avialble (using a pre-processor flag)